purchase_stock_picking_return_invoicing/models/purchase_order.py

```diff
@@ -127,15 +127,35 @@ def _compute_qty_refunded(self):
         'move_ids.returned_move_ids.state',
     )
     def _compute_qty_returned(self):
+        """Made through read_group for not impacting in performance."""
+        ProductUom = self.env['product.uom']
+        StockMove = self.env['stock.move']
+        groups = StockMove.read_group(
+            [('origin_returned_move_id.purchase_line_id', 'in', self.ids),
+             ('state', '=', 'done'),
+             ('location_id.usage', '!=', 'supplier')],
+            ['origin_returned_move_id', 'product_uom_qty', 'product_uom'],
+            ['origin_returned_move_id', 'product_uom'], lazy=False,
+        )
+        p = self._prefetch
+        # load all moves records at once on first access
+        origin_move_ids = set(g['origin_returned_move_id'][0] for g in groups)
+        StockMove.browse(origin_move_ids).with_prefetch(p)
+        # load all UoM records at once on first access
+        uom_ids = set(g['product_uom'][0] for g in groups)
+        ProductUom.browse(list(uom_ids)).with_prefetch(p)
+        line_qtys = {}
+        for g in groups:
+            uom = ProductUom.browse(g["product_uom"][0]).with_prefetch(p)
+            move = StockMove.browse(
+                g['origin_returned_move_id'][0]
+            ).with_prefetch(p)
+            line_qtys.setdefault(move.purchase_line_id.id, 0)
+            line_qtys[move.purchase_line_id.id] += uom._compute_quantity(
+                g["product_uom_qty"], move.purchase_line_id.product_uom,
+            )
         for line in self:
-            qty = 0.0
-            moves = line.mapped('move_ids.returned_move_ids')
-            for move in moves.filtered(lambda x: x.state == 'done'):
-                if move.location_id.usage != 'supplier':
-                    qty += move.product_uom._compute_quantity(
-                        move.product_uom_qty, line.product_uom,
-                    )
-            line.qty_returned = qty
+            line.qty_returned = line_qtys.get(line.id, 0)
```

Comment:
You can avoid use this extra loop if you use in the main loop:
`move.purchase_line_id['qty_returned']  += value`

since that is sharing the same object

Review the following example:
 - https://github.com/odoo/odoo/pull/25280/commits/0f237f129e369b0edbab83209ce6bfd58436060d

Reply:
I'm not sure that's a good example, as dependent values are going to be recomputed each time you assign the variable. Another option can be to enclose everything on a `with self.env.norecompute():` clause, but current option is not bad at all.

Reply:
if you uses “move.purchase_line_id[“qty_returned”] += qty” this is not re-comouted
It is just updating the cache value and you avoid an extra loop-for unneeded.

It just requires at start a self.update(qty_returned: 0) for the case where one record is not found.